<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Power calculation for a joint analysis of a two-stage case control design
for SNP data."><title>Power calculation for a joint analysis of a two-stage case control design
for SNP data — cats • cats</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Power calculation for a joint analysis of a two-stage case control design
for SNP data — cats"><meta property="og:description" content="Power calculation for a joint analysis of a two-stage case control design
for SNP data."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">cats</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/cats.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/nanxstats/cats/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Power calculation for a joint analysis of a two-stage case control design
for SNP data</h1>
      <small class="dont-index">Source: <a href="https://github.com/nanxstats/cats/blob/HEAD/R/cats.R" class="external-link"><code>R/cats.R</code></a></small>
      <div class="d-none name"><code>cats.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Power calculation for a joint analysis of a two-stage case control design
for SNP data.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">cats</span><span class="op">(</span>
  freq <span class="op">=</span> <span class="fl">0.5</span>,
  freq2 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,
  ncases <span class="op">=</span> <span class="fl">500</span>,
  ncontrols <span class="op">=</span> <span class="fl">500</span>,
  ncases2 <span class="op">=</span> <span class="fl">500</span>,
  ncontrols2 <span class="op">=</span> <span class="fl">500</span>,
  risk <span class="op">=</span> <span class="fl">1.5</span>,
  risk2 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,
  pisamples <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,
  prevalence <span class="op">=</span> <span class="fl">0.1</span>,
  prevalence2 <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,
  additive <span class="op">=</span> <span class="fl">0</span>,
  recessive <span class="op">=</span> <span class="fl">0</span>,
  dominant <span class="op">=</span> <span class="fl">0</span>,
  multiplicative <span class="op">=</span> <span class="fl">1</span>,
  alpha <span class="op">=</span> <span class="fl">1e-07</span>,
  pimarkers <span class="op">=</span> <span class="fl">0.00316</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>freq</dt>
<dd><p>numeric. The minor allele frequency (MAF) in the first stage</p></dd>
<dt>freq2</dt>
<dd><p>numeric. The MAF in the second stage, Optional, if -1 the same
value as for the first stage is given</p></dd>
<dt>ncases</dt>
<dd><p>integer. The number of cases in the first stage</p></dd>
<dt>ncontrols</dt>
<dd><p>integer. The number of controls in the first stage</p></dd>
<dt>ncases2</dt>
<dd><p>integer. The number of cases in the second stage</p></dd>
<dt>ncontrols2</dt>
<dd><p>integer. The number of controls in the second stage</p></dd>
<dt>risk</dt>
<dd><p>numeric. The relative risk in the first stage</p></dd>
<dt>risk2</dt>
<dd><p>numeric. The relative risk in the second stage, Optional, if -1
the same value as for the first stage is given</p></dd>
<dt>pisamples</dt>
<dd><p>numeric. The weights used for the joint statistic.
Optional. see details</p></dd>
<dt>prevalence</dt>
<dd><p>numeric. The prevalence of the disease in the population
for the first stage</p></dd>
<dt>prevalence2</dt>
<dd><p>numeric. The prevalence of the disease in the population
for the second stag, Optional, if -1 the same value as for the first stage
is given</p></dd>
<dt>additive</dt>
<dd><p>boolean. if 1 an additive model is assumed</p></dd>
<dt>recessive</dt>
<dd><p>boolean. if 1 a recessive model is assumed</p></dd>
<dt>dominant</dt>
<dd><p>boolean. if 1 a dominant model is assumed</p></dd>
<dt>multiplicative</dt>
<dd><p>boolean. if 1 a multiplicative model is assumed</p></dd>
<dt>alpha</dt>
<dd><p>numeric. The significance threshold. Often the a threshold of
0.05 divided by the number of markers is chosen</p></dd>
<dt>pimarkers</dt>
<dd><p>numeric. The fraction of markers genotyped in the second
stage</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <dl><dt>P.one.study</dt>
<dd><p>The power if only one study was performed, NB!
This is only a valid estimate if the relative risk and allele frequency is
the same for both stages</p></dd>
<dt>P.first.stage</dt>
<dd><p>The power for a marker to
proceed the the second stage</p></dd>
<dt>P.rep.study </dt>
<dd><p>The power of the study if
based on replication and not a joint analysis</p></dd>
<dt>P.joint.min </dt>
<dd><p>The power
of the joint analysis tp detect at least one susceptibility SNP assuming
that five susceptibility SNPs exits</p></dd>
<dt>P.joint </dt>
<dd><p>The power of the joint
analysis</p></dd>
<dt>pi </dt>
<dd><p>The weight used to calculate the joint statistic</p></dd>
<dt>T.one.study</dt>
<dd><p>Recommended thresholds for a one-stage study</p></dd>
<dt>T.first.stage </dt>
<dd><p>Recommended thresholds for the first stage in
two-stage study</p></dd>
<dt>T.second.stage.rep </dt>
<dd><p>Recommended thresholds for the
second stage in replication analysis</p></dd>
<dt>T.second.stage.joint
</dt>
<dd><p>Recommended thresholds for the second stage in a joint analysis</p></dd>
<dt>E.Disease.freq.cases1 </dt>
<dd><p>The expected disease allele frequency in stage
1 for cases</p></dd>
<dt>E.Disease.freq.controls1 </dt>
<dd><p>The expected disease allele
frequency in stage 1 for controls</p></dd>
<dt>E.Disease.freq.cases2 </dt>
<dd><p>The
expected disease allele frequency in stage 2 for cases</p></dd>
<dt>E.Disease.freq.controls2 </dt>
<dd><p>The expected disease allele frequency in
stage 2 for controls</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>These power analysis are based on Skol et al. 2006, But are generized so
that the ratio between cases and controls may vary between stages. Also the
allele frequencies, disease prevalence and relative risk are also allowed to
vary. The joint statistic $z_joint=z_1\sqrt\pi+z_2\sqrt1-\pi$ where $z_1$ is
the z-score for the first stage and the weight $\pi$ is calculated as
$\pi=1/var(\hatp'_1-\hatp_1)*(1/var(\hatp'_1-\hatp_1)+1/var(\hatp'_2-\hatp_2))^-1$,
where $\hatp'_1$ is the estimate of the allele frequency of the cases in the
first stage. This is consistent with Skol et al 2006 when the ratios of
cases and controls are the same in both stages. When this is not the case
the weight $\pi$ may vary slightly with different allele frequencies and
different relative risks. For power calculations I would recommend
calculating the weight at a likely scenario where there is about 80-90%
power and fixing the weights at other scenarios (and the testing of the real
data) to this weight. This can be done by assigning pisample to a value. In
practice this will hardly affect the power.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Skol AD, Scott LJ, Abecasis GR, Boehnke M: Joint analysis is
more efficient than replication-based analysis for two-stage genome-wide
association studies. Nat Genet 38: 209-213, 2006.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><a href="http://www.sph.umich.edu/csg/abecasis/CaTS/" class="external-link">http://www.sph.umich.edu/csg/abecasis/CaTS/</a></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Anders Albrechtsen</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="co"># calculate the power under a multiplicative model using a two stage design</span></span>
<span class="r-in"><span class="co"># and assuming a relative risk of 1.5</span></span>
<span class="r-in"><span class="fu">cats</span><span class="op">(</span></span>
<span class="r-in">  freq <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span class="r-in">  ncases <span class="op">=</span> <span class="fl">500</span>, ncases2 <span class="op">=</span> <span class="fl">500</span>,</span>
<span class="r-in">  ncontrols <span class="op">=</span> <span class="fl">1000</span>, ncontrols2 <span class="op">=</span> <span class="fl">1000</span>,</span>
<span class="r-in">  risk <span class="op">=</span> <span class="fl">1.5</span>, multiplicative <span class="op">=</span> <span class="fl">1</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Expected Power is;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    For a one-stage study = 0.94</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       For first stage in two-stage study = 0.972</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> For second stage in replication analysis = 0.784</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>     For second stage in a joint analysis = 0.929</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                       pi = 0.5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"></span>
<span class="r-in"><span class="va">power.J</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span class="r-in"><span class="va">power.R</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span class="r-in"><span class="va">power.O</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span class="r-in"><span class="va">RR</span> <span class="op">&lt;-</span> <span class="fl">23</span><span class="op">:</span><span class="fl">32</span> <span class="op">/</span> <span class="fl">20</span></span>
<span class="r-in"><span class="kw">for</span> <span class="op">(</span><span class="va">tal</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">RR</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span class="r-in">  <span class="va">temp</span> <span class="op">&lt;-</span> <span class="fu">cats</span><span class="op">(</span>risk <span class="op">=</span> <span class="va">RR</span><span class="op">[</span><span class="va">tal</span><span class="op">]</span><span class="op">)</span></span>
<span class="r-in">  <span class="va">power.J</span><span class="op">[</span><span class="va">tal</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">temp</span><span class="op">$</span><span class="va">P.joint</span></span>
<span class="r-in">  <span class="va">power.R</span><span class="op">[</span><span class="va">tal</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">temp</span><span class="op">$</span><span class="va">P.rep.study</span></span>
<span class="r-in">  <span class="va">power.O</span><span class="op">[</span><span class="va">tal</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">temp</span><span class="op">$</span><span class="va">P.one.study</span></span>
<span class="r-in"><span class="op">}</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">RR</span>, <span class="va">power.J</span>, type <span class="op">=</span> <span class="st">"b"</span>, lwd <span class="op">=</span> <span class="fl">2</span>, ylab <span class="op">=</span> <span class="st">"Power"</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html" class="external-link">lines</a></span><span class="op">(</span><span class="va">RR</span>, <span class="va">power.R</span>, lwd <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="fl">2</span>, type <span class="op">=</span> <span class="st">"b"</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html" class="external-link">lines</a></span><span class="op">(</span><span class="va">RR</span>, <span class="va">power.O</span>, lwd <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="fl">3</span>, type <span class="op">=</span> <span class="st">"b"</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">1.4</span>, <span class="fl">0.4</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span class="r-in">  <span class="st">"joint analysis"</span>, <span class="st">"replication design"</span>,</span>
<span class="r-in">  <span class="st">"one stage design"</span></span>
<span class="r-in"><span class="op">)</span>, col <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, lwd <span class="op">=</span> <span class="fl">3</span>, bty <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></span>
<span class="r-plt img"><img src="cats-1.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Anders Albrechtsen, Nan Xiao.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

    </footer></div>

  

  

  </body></html>

